### ISSUE TITLE
KeyError: 'maskIRAVEN is not in the dataset registry'

### ISSUE THREAD

**Issue by zunction** [open-mmlab/mmclassification#625]
### Checklist
- I have searched related issues but cannot get the expected help. yes
- I have read related documents and don't know what to do. yes

### Describe the question you meet

Encounter this error message:
```
Traceback (most recent call last):
  File "tools/train-perception-maskpositions.py", line 206, in <module>
    main()
  File "tools/train-perception-maskpositions.py", line 181, in main
    datasets = [build_dataset(cfg.data.train)]
  File "/root/anaconda3/envs/perceptry/lib/python3.7/site-packages/mmcls/datasets/builder.py", line 39, in build_dataset
    dataset = build_from_cfg(cfg, DATASETS, default_args)
  File "/root/anaconda3/envs/perceptry/lib/python3.7/site-packages/mmcv/utils/registry.py", line 45, in build_from_cfg
    f'{obj_type} is not in the {registry.name} registry')
KeyError: 'maskIRAVEN is not in the dataset registry'

```

### Post related information
1. The output of `pip list | grep "mmcv\|mmcls\|^torch"`
```
mmcls                         0.18.0
mmcv-full                     1.4.0
torch                         1.7.0
torchvision                   0.8.1
```
2. Your config file if you modified it or created a new one.
```python
_base_ = [
    '../_base_/models/resnet18.py', '../_base_/datasets/imagenet_bs32.py',
    '../_base_/schedules/imagenet_bs256.py', '../_base_/default_runtime.py'
]

dataset_type = 'maskIRAVEN'
classes = ('(0.16, 0.16)', '(0.16, 0.5)', '(0.16, 0.83)', '(0.25, 0.25)', '(0.25, 0.5)', 
            '(0.25, 0.75)', '(0.42, 0.42)', '(0.42, 0.58)', '(0.5, 0.16)', '(0.5, 0.25)', 
            '(0.5, 0.5)', '(0.5, 0.75)', '(0.5, 0.83)', '(0.58, 0.42)', '(0.58, 0.58)', 
            '(0.75, 0.25)', '(0.75, 0.5)', '(0.75, 0.75)', '(0.83, 0.16)', '(0.83, 0.5)', 
            '(0.83, 0.83)')

data = dict(
    samples_per_gpu=16,
    workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        data_prefix = './dataset/midformat/positions-60-20-20/train',
         ),
    val=dict(
        type=dataset_type,
        data_prefix = './dataset/midformat/positions-60-20-20/val',
        ann_file='val-positions.txt',
 
    ),
    test=dict(
        type=dataset_type,
        data_prefix = './dataset/midformat/positions-60-20-20/test',
        ann_file='test-positions.txt',
       
    )
)
data_root = './dataset/midformat/positions-60-20-20'
```
3. Your train log file if you meet the problem during training.
Training cannot start
4. Other code you modified in the `mmcls` folder.

In the `mmcls/dataset/` directory:

- added `maskiraven.py`
```
import mmcv
import numpy as np

from .base_dataset import BaseDataset
from .builder import DATASETS

@DATASETS.register_module()
class maskIRAVEN(BaseDataset):

    def load_annotations(self):
        assert isinstance(self.ann_file, str)

        data_infos = []
        with open(self.ann_file) as f:
            samples = [x.strip().split(' ') for x in f.readlines()]
            for filename, gt_label in samples:
                info = {'img_prefix': self.data_prefix}
                info['img_info'] = {'filename': filename}
                info['gt_label'] = np.array(gt_label, dtype=np.int64)
                data_infos.append(info)
            return data_infos
```
- edited `__init__.py`
```
# Copyright (c) OpenMMLab. All rights reserved.
from .base_dataset import BaseDataset
from .builder import (DATASETS, PIPELINES, SAMPLERS, build_dataloader,
                      build_dataset, build_sampler)
from .cifar import CIFAR10, CIFAR100
from .dataset_wrappers import (ClassBalancedDataset, ConcatDataset,
                               RepeatDataset)
from .imagenet import ImageNet
from .imagenet21k import ImageNet21k
from .mnist import MNIST, FashionMNIST
from .multi_label import MultiLabelDataset
from .samplers import DistributedSampler, RepeatAugSampler
from .voc import VOC

from .maskiraven import maskIRAVEN

__all__ = [
    'BaseDataset', 'ImageNet', 'CIFAR10', 'CIFAR100', 'maskIRAVEN', 'MNIST', 'FashionMNIST',
    'VOC', 'MultiLabelDataset', 'build_dataloader', 'build_dataset',
    'DistributedSampler', 'ConcatDataset', 'RepeatDataset',
    'ClassBalancedDataset', 'DATASETS', 'PIPELINES', 'ImageNet21k', 'SAMPLERS',
    'build_sampler', 'RepeatAugSampler'
]

```

**Comment by mzr1996** [comment]
Please uninstall your mmcls package and install it with the [install tutorials to develop based on mmcls](https://mmclassification.readthedocs.io/en/latest/install.html#develop-based-on-mmclassification)

**Comment by zunction** [comment]
> Please uninstall your mmcls package and install it with the [install tutorials to develop based on mmcls](https://mmclassification.readthedocs.io/en/latest/install.html#develop-based-on-mmclassification)

Hi, I previously had installed mmdetection based on the installation notes, after which I installed mmcls using mim. Also according to the link which you have provided, I have the required versions of mmcv.

Would you be able to explain what is causing the error? And, if I require mmcls and mmdet installations in a single conda environment, is that possible?

Many thanks

**Comment by mzr1996** [comment]
> 

The problem is you installed `mmcls` in the package manager's folder, in your case, `/root/anaconda3/envs/perceptry/lib/python3.7/site-packages/mmcls`.

And then, you added a dataset class in your cloned mmclassification repo, but Python still tries to find it in `/root/anaconda3/envs/perceptry/lib/python3.7/site-packages/mmcls`.

Therefore, to solve the problem, you need to uninstall the `mmcls` with `mim uninstall mmcls` and then install it again at your cloned repo with `mim install -e .`

**Comment by zunction** [comment]
Your explanation helped me understand me error. I have managed to resolve it. 

Two things which I did, in case someone in future encounters the same issue as me:
- uninstall `mmcls` and re-install `mmcls`
-  check that `data_prefix` in my configuration file is pointing to the right directory. In case you cannot tell if its is pointing correctly, use absolute path rather than relative path to the directories.

Thanks @mzr1996 for your help once again ◡̈
